src/Type/Php/ArraySumFunctionDynamicReturnTypeExtension.php

```diff
@@ -35,21 +45,85 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 			return new ConstantIntegerType(0);
 		}
 
-		$intUnionFloat = new UnionType([new IntegerType(), new FloatType()]);
+		$constantArray = $arrayType->getConstantArrays()[0] ?? null;
+
+		if ($constantArray !== null) {
+			$nonScalarTypes = [];
+			$scalarValues = [];
 
-		if ($arrayType->isIterableAtLeastOnce()->yes()) {
-			if ($intUnionFloat->isSuperTypeOf($itemType)->yes()) {
-				return $itemType;
+			$types = $constantArray->getValueTypes();
+
+			foreach ($types as $type) {
+				$scalar = $type->getConstantScalarValues();
```

Comment:
Hi, I don't really love that this PR is working with specific scalar values and calls `array_sum` internally. It misses out on various properties of the type system, like working with IntegerRangeType.

I'd much rather be if for ConstantArrayType objects, like `[$one, $two, $three]`, it tried to do `$scope->getType(new Plus($one, new Plus($two, $three)))`. Because Plus expr handling in MutatingScope already knows how to work with al possible Types, we can just call it here like that.

Because ConstantArrayType contains Type and not Expr, you can take advantage of TypeExpr virtual node to build an Expr out of Type instances.